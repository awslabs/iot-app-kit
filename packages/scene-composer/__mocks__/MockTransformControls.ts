export class MockTransformControls {
  static instance;

  object: any = {};
  showY = true;
  flipY = false;
  camera;
  domElement;

  constructor(camera, domElement) {
    this.camera = camera;
    this.domElement = domElement;

    MockTransformControls.instance = this;
  }

  static addEventListener = vi.fn();
  static removeEventListener = vi.fn();
  static setMode = vi.fn();
  static attach = vi.fn();
  static detach = vi.fn();

  addEventListener = MockTransformControls.addEventListener;
  removeEventListener = MockTransformControls.removeEventListener;
  setMode = MockTransformControls.setMode;
  attach = MockTransformControls.attach;
  detach = MockTransformControls.detach;
  traverse = vi.fn();
}
