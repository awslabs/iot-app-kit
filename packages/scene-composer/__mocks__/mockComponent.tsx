import React from 'react';

export default (componentName) => {
  // eslint-disable-next-line react/prop-types
  const component = ({ children, ...props }) => {
    const snapshotableProps = {};
    const renderProps = [];

    Object.entries(props).forEach(([key, value]) => {
      try {
        // Covers edge case where stringify fails due to circular object graphs
        if (typeof value === 'object') {
          if (React.isValidElement(value)) {
            renderProps[key] = value;
          } else if (key === 'style') {
            snapshotableProps[key] = value;
          } else {
            snapshotableProps[key] = JSON.stringify(value);
          }
        } else {
          snapshotableProps[key] = value;
        }
      } catch (e) {
        snapshotableProps[key] = value;
      }
    });

    return (
      <div data-mocked={componentName} {...snapshotableProps}>
        {Object.entries(renderProps).map(([key, value]) => (
          <div key={key}>{value}</div>
        ))}
        {children}
      </div>
    );
  };

   component.displayName = `Mock${componentName}`;
   return component
};
