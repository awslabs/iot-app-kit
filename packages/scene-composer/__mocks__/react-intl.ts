const strings = require('../translations/IotAppKitSceneComposer.en_US.json');

const reactIntl = jest.requireActual('react-intl');
type Message = {
  note: string;
  text: string;
};

const messages = Object.entries(strings).reduce((acc, [key, msg]: [key: string, msg: any]) => {
  acc[key] = msg.text;
  return acc;
}, {});

const intl = new reactIntl.createIntl({
  locale: 'en',
  messages
});

module.exports = {
  ...reactIntl,
  useIntl: jest.fn(() => intl),
};
