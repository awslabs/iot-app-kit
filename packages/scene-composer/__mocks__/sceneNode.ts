import { ISceneNodeInternal } from "../src/store";

export const defaultNode: ISceneNodeInternal = {
    ref: 'node-ref',
    name: 'default-node',
    transform: {
        position: [0, 0, 0],
        rotation: [0, 0, 0],
        scale: [1, 1, 1]
    },
    transformConstraint: {},
    childRefs: [],
    components: [],
    properties: {},
};
