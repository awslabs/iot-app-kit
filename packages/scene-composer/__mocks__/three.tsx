/**
 * Mock implementation of R3F (React Three Fiber) components with the string name for a clean snapshot test.
 */
module.exports = {
  ...jest.requireActual('three'),
  audioListener: 'audioListener',
  positionalAudio: 'positionalAudio',

  mesh: 'mesh',
  instancedMesh: 'instancedMesh',
  scene: 'scene',
  sprite: 'sprite',
  lOD: 'lOD',
  skinnedMesh: 'skinnedMesh',
  skeleton: 'skeleton',
  bone: 'bone',
  lineSegments: 'lineSegments',
  lineLoop: 'lineLoop',
  points: 'points',
  group: 'group',
  immediateRenderObject: 'immediateRenderObject',

  // cameras
  camera: 'camera',
  perspectiveCamera: 'perspectiveCamera',
  orthographicCamera: 'orthographicCamera',
  cubeCamera: 'cubeCamera',
  arrayCamera: 'arrayCamera',

  // geometry
  instancedBufferGeometry: 'instancedBufferGeometry',
  bufferGeometry: 'bufferGeometry',
  boxBufferGeometry: 'boxBufferGeometry',
  circleBufferGeometry: 'circleBufferGeometry',
  coneBufferGeometry: 'coneBufferGeometry',
  cylinderBufferGeometry: 'cylinderBufferGeometry',
  dodecahedronBufferGeometry: 'dodecahedronBufferGeometry',
  extrudeBufferGeometry: 'extrudeBufferGeometry',
  icosahedronBufferGeometry: 'icosahedronBufferGeometry',
  latheBufferGeometry: 'latheBufferGeometry',
  octahedronBufferGeometry: 'octahedronBufferGeometry',
  parametricBufferGeometry: 'parametricBufferGeometry',
  planeBufferGeometry: 'planeBufferGeometry',
  polyhedronBufferGeometry: 'polyhedronBufferGeometry',
  ringBufferGeometry: 'ringBufferGeometry',
  shapeBufferGeometry: 'shapeBufferGeometry',
  sphereBufferGeometry: 'sphereBufferGeometry',
  tetrahedronBufferGeometry: 'tetrahedronBufferGeometry',
  textBufferGeometry: 'textBufferGeometry',
  torusBufferGeometry: 'torusBufferGeometry',
  torusKnotBufferGeometry: 'torusKnotBufferGeometry',
  tubeBufferGeometry: 'tubeBufferGeometry',
  wireframeGeometry: 'wireframeGeometry',
  parametricGeometry: 'parametricGeometry',
  tetrahedronGeometry: 'tetrahedronGeometry',
  octahedronGeometry: 'octahedronGeometry',
  icosahedronGeometry: 'icosahedronGeometry',
  dodecahedronGeometry: 'dodecahedronGeometry',
  polyhedronGeometry: 'polyhedronGeometry',
  tubeGeometry: 'tubeGeometry',
  torusKnotGeometry: 'torusKnotGeometry',
  torusGeometry: 'torusGeometry',
  textGeometry: 'textGeometry',
  sphereGeometry: 'sphereGeometry',
  ringGeometry: 'ringGeometry',
  planeGeometry: 'planeGeometry',
  latheGeometry: 'latheGeometry',
  shapeGeometry: 'shapeGeometry',
  extrudeGeometry: 'extrudeGeometry',
  edgesGeometry: 'edgesGeometry',
  coneGeometry: 'coneGeometry',
  cylinderGeometry: 'cylinderGeometry',
  circleGeometry: 'circleGeometry',
  boxGeometry: 'boxGeometry',

  // materials
  material: 'material',
  shadowMaterial: 'shadowMaterial',
  spriteMaterial: 'spriteMaterial',
  rawShaderMaterial: 'rawShaderMaterial',
  shaderMaterial: 'shaderMaterial',
  pointsMaterial: 'pointsMaterial',
  meshPhysicalMaterial: 'meshPhysicalMaterial',
  meshStandardMaterial: 'meshStandardMaterial',
  meshPhongMaterial: 'meshPhongMaterial',
  meshToonMaterial: 'meshToonMaterial',
  meshNormalMaterial: 'meshNormalMaterial',
  meshLambertMaterial: 'meshLambertMaterial',
  meshDepthMaterial: 'meshDepthMaterial',
  meshDistanceMaterial: 'meshDistanceMaterial',
  meshBasicMaterial: 'meshBasicMaterial',
  meshMatcapMaterial: 'meshMatcapMaterial',
  lineDashedMaterial: 'lineDashedMaterial',
  lineBasicMaterial: 'lineBasicMaterial',

  // primitive
  primitive: 'primitive',

  // lights and other
  light: 'light',
  spotLightShadow: 'spotLightShadow',
  spotLight: 'spotLight',
  pointLight: 'pointLight',
  rectAreaLight: 'rectAreaLight',
  hemisphereLight: 'hemisphereLight',
  directionalLightShadow: 'directionalLightShadow',
  directionalLight: 'directionalLight',
  ambientLight: 'ambientLight',
  lightShadow: 'lightShadow',
  ambientLightProbe: 'ambientLightProbe',
  hemisphereLightProbe: 'hemisphereLightProbe',
  lightProbe: 'lightProbe',

  // helpers
  spotLightHelper: 'spotLightHelper',
  skeletonHelper: 'skeletonHelper',
  pointLightHelper: 'pointLightHelper',
  hemisphereLightHelper: 'hemisphereLightHelper',
  gridHelper: 'gridHelper',
  polarGridHelper: 'polarGridHelper',
  directionalLightHelper: 'directionalLightHelper',
  cameraHelper: 'cameraHelper',
  boxHelper: 'boxHelper',
  box3Helper: 'box3Helper',
  planeHelper: 'planeHelper',
  arrowHelper: 'arrowHelper',
  axesHelper: 'axesHelper',

  // textures
  texture: 'texture',
  videoTexture: 'videoTexture',
  dataTexture: 'dataTexture',
  dataTexture3D: 'dataTexture3D',
  compressedTexture: 'compressedTexture',
  cubeTexture: 'cubeTexture',
  canvasTexture: 'canvasTexture',
  depthTexture: 'depthTexture',

  // misc
  raycaster: 'raycaster',
  vector2: 'vector2',
  vector3: 'vector3',
  vector4: 'vector4',
  euler: 'euler',
  matrix3: 'matrix3',
  matrix4: 'matrix4',
  quaternion: 'quaternion',
  bufferAttribute: 'bufferAttribute',
  instancedBufferAttribute: 'instancedBufferAttribute',
  color: 'color',
  fog: 'fog',
  fogExp2: 'fogExp2',
  shape: 'shape',
};
