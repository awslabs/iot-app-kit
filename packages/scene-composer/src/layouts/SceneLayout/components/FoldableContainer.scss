@use '@cloudscape-design/design-tokens/index.scss' as awsui;

.tm-wrapper-left,
.tm-wrapper-right {
  display: flex;
  align-items: stretch;
  height: 100%;
}

.tm-wrapper-left { flex-direction: row; }
.tm-wrapper-right { flex-direction: row-reverse; }
.tm-content { flex: 1; }

.tm-handle {
  flex: none;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  border-left: 1px solid awsui.$color-border-divider-default;
  border-right: 1px solid awsui.$color-border-divider-default;

  &:hover {
    background-color: awsui.$color-background-dropdown-item-hover;
  }
}
