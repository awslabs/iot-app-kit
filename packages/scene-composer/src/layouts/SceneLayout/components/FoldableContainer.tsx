import React, { useEffect, useState } from 'react';
import { Icon, IconProps } from '@awsui/components-react';

import { Direction } from './utils';
import './FoldableContainer.scss';

type FoldableContainerProps = React.PropsWithChildren<{
  direction: Direction;
  open: boolean;
  setIsOpen: (boolean) => void;
}>;

const FoldableContainer: React.FC<FoldableContainerProps> = ({
  direction,
  children,
  open,
  setIsOpen,
}: FoldableContainerProps) => {
  const [fold, setFold] = useState(!open);
  const isLeft = direction === 'Left';
  const isRight = !isLeft;

  const iconName: IconProps.Name = (fold && isLeft) || (!fold && isRight) ? 'angle-right' : 'angle-left';

  const wrapper = `tm-wrapper-${isLeft ? 'left' : 'right'}`;

  useEffect(() => {
    setIsOpen(!fold);
  }, [fold]);

  return (
    <div className={wrapper}>
      {!fold && <div className='tm-content'>{children}</div>}
      <div className='tm-handle' data-testid='handle' onClick={() => setFold(!fold)}>
        <Icon name={iconName} size='small' variant='normal' />
      </div>
    </div>
  );
};

export default FoldableContainer;
