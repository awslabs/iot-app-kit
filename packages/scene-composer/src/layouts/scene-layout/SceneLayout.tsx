import React, { FC, Fragment, ReactNode, Suspense, useCallback, useContext } from 'react';
import { useIntl } from 'react-intl';
import styled, { ThemeContext } from 'styled-components';
import { Canvas, ThreeEvent } from '@react-three/fiber';
import { useContextBridge } from '@react-three/drei/core/useContextBridge';
import { isEmpty } from 'lodash';

import LoggingContext from '../../logger/react-logger/contexts/logging';
import MessageModal from '../../components/MessageModal';
import { MenuBar } from '../../components/MenuBar';
import { StaticLayout } from '../../components/StaticLayout';
import { WebGLCanvasManager } from '../../components/WebGLCanvasManager';
import { FloatingToolbar } from '../../components/toolbars';
import {
  SceneHierarchyPanel,
  SceneNodeInspectorPanel,
  SceneRulesPanel,
  SettingsPanel,
  TopBar,
} from '../../components/panels';
import { sceneComposerIdContext } from '../../sceneComposerIdContext';
import { useSceneDocument, useStore } from '../../store';
import { KnownComponentType, KnownSceneProperty } from '../../interfaces';
import { PoweredByMatterport } from '../../assets/svgs';

import LeftPanel from './components/LeftPanel';
import RightPanel from './components/RightPanel';

const UnselectableCanvas = styled(Canvas)`
  user-select: none;
  background: ${({ theme }) => {
    return theme.canvasBackground;
  }};
  z-index: 0;
`;

interface SceneLayoutProps {
  isViewing: boolean;
  onPointerMissed: (event: ThreeEvent<PointerEvent>) => void;
  LoadingView: ReactNode;
  showMessageModal: boolean;
}

const SceneLayout: FC<SceneLayoutProps> = ({ isViewing, onPointerMissed, LoadingView = null, showMessageModal }) => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const valueDataBindingProvider = useStore(sceneComposerId)((state) => state.getEditorConfig)()
    .valueDataBindingProvider;
  const getComponentRefByType = useStore(sceneComposerId)((state) => state.getComponentRefByType);
  const componentNodeMap = useStore(sceneComposerId)((state) => state.document.componentNodeMap);
  const ContextBridge = useContextBridge(LoggingContext, sceneComposerIdContext, ThemeContext);
  const intl = useIntl();
  const { sceneLoaded, getSceneProperty } = useSceneDocument(sceneComposerId);
  const matterportModelId: string = getSceneProperty(KnownSceneProperty.MatterportModelId);

  const leftPanelEditModeProps = {
    [intl.formatMessage({ defaultMessage: 'Hierarchy', description: 'Panel Tab title' })]: <SceneHierarchyPanel />,
    [intl.formatMessage({ defaultMessage: 'Rules', description: 'Panel Tab title' })]: <SceneRulesPanel />,
    [intl.formatMessage({ defaultMessage: 'Settings', description: 'Panel Tab title' })]: (
      <SettingsPanel valueDataBindingProvider={valueDataBindingProvider} />
    ),
  };
  const leftPanelViewModeProps = {
    [intl.formatMessage({ defaultMessage: 'Hierarchy', description: 'Panel Tab title' })]: <SceneHierarchyPanel />,
  };
  const rightPanelProps = {
    [intl.formatMessage({ defaultMessage: 'Inspector', description: 'Panel Tab title' })]: <SceneNodeInspectorPanel />,
  };
  const leftPanel = <LeftPanel {...leftPanelEditModeProps} />;

  const viewingModeLeftPanel = <LeftPanel {...leftPanelViewModeProps} />;

  const rightPanel = <RightPanel {...rightPanelProps} />;

  const showTopBar = useCallback(() => {
    return !isEmpty(getComponentRefByType(KnownComponentType.MotionIndicator));
  }, [componentNodeMap]);

  return (
    <StaticLayout
      mainContent={
        <Fragment>
          <FloatingToolbar isViewing={isViewing} />
          {matterportModelId && <PoweredByMatterport matterportModelId={matterportModelId} />}
          <UnselectableCanvas shadows dpr={window.devicePixelRatio} onPointerMissed={onPointerMissed}>
            <ContextBridge>
              <Suspense fallback={LoadingView}>{!sceneLoaded ? null : <WebGLCanvasManager />}</Suspense>
            </ContextBridge>
          </UnselectableCanvas>
        </Fragment>
      }
      showModal={showMessageModal}
      modalContent={<MessageModal />}
      header={!isViewing && <MenuBar />}
      leftPanel={isViewing ? viewingModeLeftPanel : leftPanel}
      rightPanel={!isViewing && rightPanel}
      topBar={showTopBar() && <TopBar />}
    />
  );
};

export default SceneLayout;
