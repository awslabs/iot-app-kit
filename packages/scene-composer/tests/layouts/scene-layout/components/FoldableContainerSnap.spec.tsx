/* eslint-disable import/first */
import renderer from 'react-test-renderer';
import React from 'react';

// eslint-disable-next-line import/order
import { mockPolaris } from '../../../__mocks__/MockPolaris';

mockPolaris();

import FoldableContainer from '../../../../src/layouts/scene-layout/components/FoldableContainer';
import { Direction } from '../../../../src/layouts/scene-layout';

describe('FoldableContainerSnap', () => {
  it('should render a left fold correctly', () => {
    const container = renderer.create(<FoldableContainer direction={Direction.Left}>My Content</FoldableContainer>);
    expect(container).toMatchSnapshot();
  });
  it('should render a right fold correctly', () => {
    const container = renderer.create(<FoldableContainer direction={Direction.Right}>My Content</FoldableContainer>);
    expect(container).toMatchSnapshot();
  });
});
