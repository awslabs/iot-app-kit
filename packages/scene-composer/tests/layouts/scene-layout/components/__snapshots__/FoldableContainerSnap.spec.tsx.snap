// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FoldableContainerSnap should render a left fold correctly 1`] = `
<div
  className="sc-bczRLJ iNflyu"
>
  <div
    className="sc-dkzDqf JleGA"
  >
    My Content
  </div>
  <div
    className="sc-hKMtZM gUSkvB"
    onClick={[Function]}
  >
    <Icon
      name="angle-left"
      size="small"
      variant="normal"
    />
  </div>
</div>
`;

exports[`FoldableContainerSnap should render a right fold correctly 1`] = `
<div
  className="sc-gsnTZi iHBFHx"
>
  <div
    className="sc-dkzDqf JleGA"
  >
    My Content
  </div>
  <div
    className="sc-hKMtZM gUSkvB"
    onClick={[Function]}
  >
    <Icon
      name="angle-right"
      size="small"
      variant="normal"
    />
  </div>
</div>
`;
